src/interactive-window/interactiveWindow.ts

```diff
@@ -558,33 +578,39 @@ export class InteractiveWindow implements IInteractiveWindowLoadable {
 
         // Multiple cells that have split our code.
         const promises = cells.map((c) => {
+            const deferred = createDeferred<void>();
+            this.setPendingCellAdd(deferred.promise);
             // Add the cell first. We don't need to wait for this part as we want to add them
             // as quickly as possible
             const notebookCellPromise = this.addNotebookCell(c, fileUri, line);
 
             // Queue up execution
             const promise = this.createExecutionPromise(notebookCellPromise, isDebug);
-            promise.catch((ex) => {
-                // If execution fails due to a failure in another cell, then log that error against the cell.
-                if (ex instanceof InteractiveCellResultError) {
-                    notebookCellPromise
-                        .then((cell) => {
-                            if (ex.cell !== cell) {
-                                this.addErrorMessage(DataScience.cellStopOnErrorMessage(), cell).then(noop, noop);
-                            }
-                        })
-                        .catch(noop);
-                } else {
-                    notebookCellPromise
-                        .then((cell) =>
-                            // If our cell result was a failure show an error
-                            this.errorHandler
-                                .getErrorMessageForDisplayInCell(ex, 'execution', this.owningResource)
-                                .then((message) => this.addErrorMessage(message, cell))
-                        )
-                        .catch(noop);
-                }
-            });
+            promise
+                .catch((ex) => {
+                    // If execution fails due to a failure in another cell, then log that error against the cell.
+                    if (ex instanceof InteractiveCellResultError) {
+                        notebookCellPromise
+                            .then((cell) => {
+                                if (ex.cell !== cell) {
+                                    this.addErrorMessage(DataScience.cellStopOnErrorMessage(), cell).then(noop, noop);
+                                }
+                            })
+                            .catch(noop);
+                    } else {
+                        notebookCellPromise
+                            .then((cell) =>
+                                // If our cell result was a failure show an error
+                                this.errorHandler
+                                    .getErrorMessageForDisplayInCell(ex, 'execution', this.owningResource)
+                                    .then((message) => this.addErrorMessage(message, cell))
+                            )
+                            .catch(noop);
+                    }
+                })
+                .finally(() => {
+                    deferred?.resolve();
+                });
             return promise;
         });
 
```

Comment:
I think what you can do is

```suggestion
const deferred = createDeferred<void>();
this.setPendingCellAdd(deferred.promise);
promises[promises.length - 1].finally(() => deferred.resolve()).ignoreErrors();
```

Reply:
I'd rather have a promise per cell add